src/connection-manager/dialer/index.ts

```diff
@@ -202,76 +217,83 @@ export class DefaultDialer implements Startable, Dialer {
-   * The dial to the first address that is successfully able to upgrade a connection
-   * will be used.
-   */
-  async _createCancellableDialTarget (peer: PeerId, options: AbortOptions): Promise<DialTarget> {
-    // Make dial target promise cancellable
-    const id = `${(parseInt(String(Math.random() * 1e9), 10)).toString()}${Date.now()}`
-    const cancellablePromise = new Promise<DialTarget>((resolve, reject) => {
-      this.pendingDialTargets.set(id, { resolve, reject })
-    })
-
-    try {
-      const dialTarget = await Promise.race([
-        this._createDialTarget(peer, options),
-        cancellablePromise
-      ])
-
-      return dialTarget
-    } finally {
-      this.pendingDialTargets.delete(id)
-    }
-  }
-
   /**
    * Creates a DialTarget. The DialTarget is used to create and track
    * the DialRequest to a given peer.
-   * If a multiaddr is received it should be the first address attempted.
+   *
+   * If a multiaddr is received it should be the only address attempted.
+   *
    * Multiaddrs not supported by the available transports will be filtered out.
    */
-  async _createDialTarget (peer: PeerId, options: AbortOptions): Promise<DialTarget> {
+  async _createDialTarget (peerIdOrMultiaddr: PeerId | Multiaddr, options: AbortOptions): Promise<DialTarget> {
     const _resolve = this._resolve.bind(this)
 
-    let addrs = await pipe(
-      await this.components.peerStore.addressBook.get(peer),
-      (source) => filter(source, async (address) => {
-        return !(await this.components.connectionGater.denyDialMultiaddr(peer, address.multiaddr))
-      }),
-      // Sort addresses so, for example, we try certified public address first
-      (source) => sort(source, this.addressSorter),
-      async function * resolve (source) {
-        for await (const a of source) {
-          yield * await _resolve(a.multiaddr, options)
-        }
-      },
+    let addrs = isMultiaddr(peerIdOrMultiaddr) ? [peerIdOrMultiaddr] : await this._loadAddresses(peerIdOrMultiaddr)
+
+    addrs = (await Promise.all(
+      addrs.map(async (ma) => await _resolve(ma, options))
+    ))
+      // flatten
+      .reduce((acc, curr) => {
+        return [
+          ...acc,
+          ...curr
+        ]
+      }, [])
       // Multiaddrs not supported by the available transports will be filtered out.
-      (source) => filter(source, (ma) => Boolean(this.components.transportManager.transportForMultiaddr(ma))),
-      (source) => map(source, (ma) => {
-        if (peer.toString() === ma.getPeerId()) {
-          return ma
-        }
-
-        return ma.encapsulate(`/p2p/${peer.toString()}`)
-      }),
-      async (source) => await all(source)
-    )
+      .filter(ma => Boolean(this.components.transportManager.transportForMultiaddr(ma)))
 
+    // deduplicate addresses
     addrs = [...new Set(addrs)]
 
     if (addrs.length > this.maxAddrsToDial) {
-      await this.components.peerStore.delete(peer)
```

Comment:
if user passes ma, we could replace records with given ma?

Reply:
Not sure, that sounds surprising.  I think it would be better for the user to respond to the error message and update the address book if they need to.